Tutorials/intro-to-QUAlibrate/calibration_utils/time_of_flight/analysis.py

```diff
@@ -0,0 +1,71 @@
+"""
+Signal processing utilities for Qualibrate nodes.
+"""
+
+from scipy.signal import savgol_filter
+import numpy as np
+
+from qualibrate import QualibrationNode
+
+from qualang_tools.units import unit
+
+# Class containing tools to help handle units and conversions.
+u = unit(coerce_to_integer=True)
+
+
+def process_raw_data(fetched_data: dict):
+    """
+    Convert raw ADC traces into volts.
+
+    Args:
+        fetched_data (dict): Dictionary of raw results (typically from the execution node).
+
+    Modifies:
+        fetched_data (in-place): All NumPy array values are converted to volts using unit conversions.
+    """
+    # Convert raw ADC traces into volts
+    for key, value in fetched_data.items():
+        if isinstance(value, np.ndarray):
+            fetched_data[key] = u.raw2volts(value)
+
+
+def fit_raw_data(fetched_data: dict, node: QualibrationNode):
+    """
+    Analyze raw ADC data to extract features like mean and pulse arrival delay.
+
+    Performs the following steps:
+    - Computes and stores the mean of each signal (under "mean_values").
+    - Mean-centers each ADC trace.
+    - Detects pulse arrival time via Savitzky-Golay filter + threshold.
+
+    Args:
+        fetched_data (dict): Dictionary of preprocessed results (in volts).
+        node (QualibrationNode): Node containing metadata (especially `node.parameters.resonators`).
+
+    Modifies:
+        fetched_data (in-place):
+            - Adds "mean_values" dict
+            - Adds "delay{i}" entries for each resonator
+    """
+    # Derive the average values
+    mean_values = {}
+    for key, value in fetched_data.items():
+        if isinstance(value, np.ndarray):
+            mean_values[key] = value.mean()
+        elif isinstance(value, (int, float)):
+            mean_values[key] = value
+    # Save the average values
+    fetched_data["mean_values"] = mean_values
```

Comment:
Here I would split the raw_data from the analyzed data in node.results, in order to better structure the dict and make it easier to explore and debug

Reply:
I put the rae data in the raw data key so then we have raw_data, mean_values and delay which I think it good :)

Reply:
@TheoLaudatQM  let me know if you disagree or resolve it :)

Reply:
What I meant is to structure node.results as:
```
    node.results["ds_raw"] = raw_data (I&Q data fetched from the OPX without additional processing)
    node.results["ds_fit"] = fitted_data (I&Q data after processing)
    node.results["fit_results"] = fitted_results (results from the analysis like tof, mean_values...)
```
Which clearly identifies the main key of the results object.
Otherwise, you are just adding data to it without identifying it, which can create confusion.

Reply:
Ok, I changed it a bit so now it does make more sense :) 
Now insode `node.results.keys()` we have:
dict_keys(['raw_data', 'processed_data', 'fitted_data', 'figures'])
while 
raw_data = raw adc data
processed_data = data in volts 
fitted_data = data from fit
figures = figures